### PR TITLE
feat(forge verify-contract): `--guess-constructor-args`

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Ref #6334

## Solution

1. Added `RpcOpts` to `VerifyArgs`. If any valid `eth_rpc_url` is found in config and `--chain` is not specified, we are using explorer for chain_id from RPC. This should make UX cleaner (ref [comment](https://github.com/foundry-rs/foundry/issues/6334#issuecomment-1873562079))
2. Added `--guess-constructor-args` parameter which uses Etherscan API and RPC provider to fetch creation code of contracts created directly with create transaction or directly with default CREATE2 deployer.
3. Added test for `--guess-constructor-args`. I wasn't able to run existing verification tests without some small updates to them (renaming `--etherscan-key` to `--etherscan-api-key` for example). Is there a chance that required env vars are not set in CI which leads to tests silently passing or I got something wrong?

The only way I can see to make `--guess-constructor-args` work for any contract is by implementing some kind of `cast run` with custom inspector which will be pretty long-running thing (some `cast run`'s are taking several minutes on my machine). So not sure if such UX is really a best way to do it. Would be happy to get some ideas on how creation code for arbitrary contract could be fetched faster.

### CODE REVIEW

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 1]
```diff
@@ -458,6 +468,56 @@ impl EtherscanVerificationProvider {
             )?;
             let encoded_args = hex::encode(encoded_args);
             return Ok(Some(encoded_args[8..].into()))
+        } else if args.guess_constructor_args {
```
Comment (mattsse) [inline crates/forge/bin/cmd/verify/etherscan/mod.rs:471]: can we move this to a standalone function?

**Review by mattsse — APPROVED** [review]
style nit,

I'll set the env vars in CI via secrets

**Comment by klkvr** [comment]
made `guess_constructor_args` a separate function

also did some more testing and it seems that in some cases `compile_file` results in different bytecode, so changed `project.compile_file` to `project.compile`, it shouldn't take long as it compiles with the same settings, and will probably just use cache all the time

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 2]
```diff
@@ -459,9 +469,70 @@ impl EtherscanVerificationProvider {
...
             return Ok(Some(encoded_args[8..].into()))
         }
+        if args.guess_constructor_args {
+            return Ok(Some(self.guess_constructor_args(args, project, config).await?))
+        }
 
         Ok(args.constructor_args.clone())
     }
+
+    async fn guess_constructor_args(
```
Comment (Evalir) [inline crates/forge/bin/cmd/verify/etherscan/mod.rs:479]: could we add some docs that explains how this works?

**Review by Evalir — APPROVED** [review]
nice! only one doc nit, pending vars

**Review by onbjerg — APPROVED** [review]
lgtm

**Pull request merged**